src/nvim/buffer.c

```diff
@@ -4149,42 +4092,32 @@ void buf_signcols_add_check(buf_T *buf, sign_entry_T *added)
     return;
   }
 
-  sign_entry_T *s;
-
-  // Get first sign for added lnum
-  for (s = added; s->se_prev && s->se_lnum == s->se_prev->se_lnum; s = s->se_prev) {}
-
-  // Count signs for lnum
-  int linesum = 1;
-  for (; s->se_next && s->se_lnum == s->se_next->se_lnum; s = s->se_next) {
-    linesum++;
-  }
-  linesum += decor_signcols(buf, &decor_state, (int)s->se_lnum - 1, (int)s->se_lnum - 1,
-                            SIGN_SHOW_MAX - linesum);
+  int signcols = decor_signcols(buf, lnum - 1, lnum - 1, SIGN_SHOW_MAX);
 
-  if (linesum > buf->b_signcols.size) {
-    buf->b_signcols.size = linesum;
-    buf->b_signcols.max = linesum;
-    buf->b_signcols.sentinel = added->se_lnum;
+  if (signcols > buf->b_signcols.size) {
+    buf->b_signcols.size = signcols;
+    buf->b_signcols.max = signcols;
+    buf->b_signcols.sentinel = lnum;
     redraw_buf_later(buf, UPD_NOT_VALID);
   }
 }
 
-int buf_signcols(buf_T *buf, int maximum)
+int buf_signcols(buf_T *buf, int max)
 {
   // The maximum can be determined from 'signcolumn' which is window scoped so
   // need to invalidate signcols if the maximum is greater than the previous
-  // maximum.
-  if (maximum > buf->b_signcols.max) {
+  // (valid) maximum.
+  if (buf->b_signcols.max && max > buf->b_signcols.max) {
```

Comment:
This was flip-flopping `b_signcols.valid` in a buffer without any signs. Causing unnecessarily looping over all marks in `decor_signcols()` each time any extmark was added. (Could try adding a regression test for this, maybe through FFI?)

Reply:
good catch. Doesn't need a specific test.